tests/src/tests/test_utils.go

```diff
@@ -0,0 +1,394 @@
+}
+
+var GroupKinds = [...]schema.CommandGroupKind{schema.BuildCommandGroupKind, schema.RunCommandGroupKind, schema.TestCommandGroupKind, schema.DebugCommandGroupKind}
+
+// Return random group kind. One of "build", "run", "test" or "debug"
+func GetRandomGroupKind() schema.CommandGroupKind {
+	return GroupKinds[GetRandomNumber(len(GroupKinds))-1]
+}
+
+// Randomly returns true or false
+func GetBinaryDecision() bool {
+	return GetRandomDecision(1, 1)
+}
+
+// Randomly returns true or false, but weighted to one or the other.
+// For example if success is set to 2 and failure to 1, true is twice as likely to be returned.
+func GetRandomDecision(success int, failure int) bool {
+	setRandSeed()
+	return rand.Intn(success+failure) > failure-1
+}
+
+// Randomly returns an integer between 1 and the number specified.
+func GetRandomNumber(max int) int {
+	setRandSeed()
+	return rand.Intn(max) + 1
+}
+
+// Return a structure used to represent a specific devfile in the tests
+func GetDevfile(fileName string) TestDevfile {
+	testDevfile := TestDevfile{}
+	testDevfile.SchemaDevFile = schema.Devfile{}
+	testDevfile.FileName = fileName
+	testDevfile.SchemaDevFile.SchemaVersion = "2.0.0"
+	testDevfile.SchemaParsed = false
+	return testDevfile
+}
+
+// Create a devifle on disk for use in the tests.
+// If useParser is true the parser is used to generate the file, otherwise "sigs.k8s.io/yaml" is used.
+// The TestDevfile structure specified contains the name of the devfile and its required content.
+func (devfile *TestDevfile) CreateDevfile(useParser bool) error {
+	var err error
+
+	fileName := devfile.FileName
+	if !strings.HasSuffix(fileName, ".yaml") {
+		fileName += ".yaml"
+	}
+
+	if useParser {
+		LogInfoMessage(fmt.Sprintf("Use Parser to write devfile %s", fileName))
+		newDevfile, err := devfileData.NewDevfileData(devfile.SchemaDevFile.SchemaVersion)
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("Creating new devfile : %v", err))
+		} else {
+			newDevfile.SetSchemaVersion(devfile.SchemaDevFile.SchemaVersion)
+
+			// add the commands to new devfile
+			for _, command := range devfile.SchemaDevFile.Commands {
+				newDevfile.AddCommands(command)
+			}
+			// add components to the new devfile
+			newDevfile.AddComponents(devfile.SchemaDevFile.Components)
+
+			ctx := devfileCtx.NewDevfileCtx(fileName)
+
+			err = ctx.SetAbsPath()
+			if err != nil {
+				LogErrorMessage(fmt.Sprintf("Setting devfile path : %v", err))
+			} else {
+				devObj := parser.DevfileObj{
+					Ctx:  ctx,
+					Data: newDevfile,
+				}
+				err = devObj.WriteYamlDevfile()
+				if err != nil {
+					LogErrorMessage(fmt.Sprintf("Writing devfile : %v", err))
+				} else {
+					devfile.SchemaParsed = false
+				}
+			}
+
+		}
+	} else {
+		LogInfoMessage(fmt.Sprintf("Marshall and write devfile %s", devfile.FileName))
+		c, err := yaml.Marshal(&(devfile.SchemaDevFile))
+
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("Marshall devfile %s : %v", devfile.FileName, err))
```

Comment:
need to return here because this `err` variable is different from the `var err error` in the first line of the func. So in the end where you `return err` it will be empty when this if block is executed. Same for the if block below this

Reply:
I fixed this to make sure the error is included in return err.